igcollect/sfp_dom.py

```diff
@@ -0,0 +1,371 @@
+SFF_A2_TX_PWR_HALRM = 24
+SFF_A2_TX_PWR_LALRM = 26
+SFF_A2_TX_PWR_HWARN = 28
+SFF_A2_TX_PWR_LWARN = 30
+SFF_A2_RX_PWR = 104
+SFF_A2_RX_PWR_HALRM = 32
+SFF_A2_RX_PWR_LALRM = 34
+SFF_A2_RX_PWR_HWARN = 36
+SFF_A2_RX_PWR_LWARN = 38
+
+
+def parse_args():
+    parser = ArgumentParser()
+    parser.add_argument('--prefix', default='sfp_dom')
+    return parser.parse_args()
+
+
+def main():
+    args = parse_args()
+    physical_interfaces = get_physical_interfaces()
+
+    for iface in physical_interfaces:
+        try:
+            data = ethtool_get_module_data(iface)
+            print_metrics(args.prefix, iface, data.metrics)
+        except (NoSFPException, UnsupportedSFPException) as exc:
+            print(exc, file=sys.stderr)
+            continue
+
+
+def get_physical_interfaces():
+    scn = '/sys/class/net'
+    physical_interfaces = []
+    for dev in listdir(scn):
+        dev_path = join(scn, dev)
+        if not (islink(dev_path) or isdir(dev_path)):
+            # Skip regular files
+            continue
+        # Physical interfaces have device as a symlink to /sys/class/device_id
+        if islink(join(scn, dev, 'device')):
+            physical_interfaces.append(dev)
+
+    return physical_interfaces
+
+
+def a2_offset_to_celsius(data, offset):
+    """
+    Returns a signed float built from a 16-bit signed int. Each LSB is
+    equivalent to an increment of 1/256 degrees Celsius
+    """
+    signed = struct.unpack_from('>h', data, SFF_A2_BASE + offset)[0]
+    return signed / 256
+
+
+def a2_offset_to_volts(data, offset):
+    return a2_offset_to_int(data, offset) / 10000
+
+
+def a2_offset_to_milliamps(data, offset):
+    return a2_offset_to_int(data, offset) / 500
+
+
+def a2_offset_to_int(data, offset):
+    """
+    Returns an integer built from a 16-bit unsigned integer obtained from
+    data, at the specified offset
+    """
+    return struct.unpack_from('>H', data, SFF_A2_BASE + offset)[0]
+
+
+def print_metrics(prefix, interface, metrics):
+    timestamp = int(time())
+    for metric, child in metrics.items():
+        for value_key, value in child.items():
+            print(
+                f'{prefix}.{interface}.{metric}.{value_key}'
+                f' {value} {timestamp}'
+            )
+
+
+def ethtool_get_module_data(interface):
+    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    name = (ctypes.c_ubyte * IFNAMSIZ)(*bytearray(str(interface).encode()))
+    ifr = ifreq()
+    ifr.ifr_name = name
+
+    # Get module info
+    em = ethtool_modinfo()
+    em.cmd = ETHTOOL_GMODULEINFO
+    ifr.ifr_data.ethtool_modinfo_ptr = ctypes.pointer(em)
+
+    try:
+        fcntl.ioctl(sock, SIOCETHTOOL, ifr)
+    except OSError as exc:
+        if exc.errno == 95:
+            raise NoSFPException(f'Interface {interface} has no SFP module') \
+                from exc
+        else:
+            raise
+
+    ee = ethtool_eeprom()
+    ee.cmd = ETHTOOL_GMODULEEEPROM
+    ee.len = em.eeprom_len
+    ee.offset = 0
```

Comment:
Is it maybe possible to make the constructor of `ethtool_eeprom` so it take the name already in it?
``` python
ifr = ethtool_eeprom(cmd=ETHTOOL_GMODULEEEPROM, len=eeprom_len, offset=0)
```

Reply:
I adjusted the code to match the example above.